### PR TITLE
Support Python 3.12

### LINKED ISSUE #472: Extend support to Python 3.12
### Description

Current blocker: https://github.com/pymc-devs/pytensor/pull/444

### PR DESCRIPTION
**Pull request by maresb**
Closes #472

### Motivation for these changes
Get the latest Python goodness

### Implementation details
Basically just 3.11→3.12

### Checklist
+ [X] Explain motivation and implementation 👆
+ [X] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [X] Link relevant issues, preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [X] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes). Note that if they don't, we will [rewrite/rebase/squash the git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) before merging.
+ [X] Are the changes covered by tests and docstrings?
+ [X] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## New features
- PyTensor now works with Python 3.12

## Bugfixes
- ...

## Documentation
- ...

## Maintenance
- ...

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -52,7 +52,7 @@ jobs:
     if: ${{ needs.changes.outputs.changes == 'true' }}
     strategy:
       matrix:
-        python-version: ["3.9", "3.10", "3.11"]
+        python-version: ["3.9", "3.11", "3.12"]
```
Comment (ricardoV94) [inline .github/workflows/test.yml:55]: I think it's fine to run code style in all versions?

Comment (maresb) [reply .github/workflows/test.yml:55]: Actually, this seems very strange to me. Why would we run pre-commit on more than one Python version? Pre-commit provides environment isolation for each hook, so it shouldn't matter from what version you run it.

Comment (ricardoV94) [reply .github/workflows/test.yml:55]: It catches stuff like syntax that is not supported in older versions

Comment (ricardoV94) [reply .github/workflows/test.yml:55]: But I'm happy with just running on the oldest and most recent versions. Either that or all. A thing in between seems odd.

**Comment by ricardoV94** [comment]
What the heck is the pre-commit complaining about?

**Comment by maresb** [comment]
Switching from #581 to #586 to see what happens

**Comment by maresb** [comment]
Beautiful... we can get mypy to pass on 3.9 or 3.12 but not both. Maybe there's a way to get mypy to ignore unused type ignores...

**Comment by maresb** [comment]
There is no `mkl-service` build for Python 3.12. This seems pretty blocking.

See:
https://github.com/conda-forge/pytensor-suite-feedstock/blob/0671c8077973cfb8062aeffd0f131981067757f9/recipe/meta.yaml#L96-L99

**Comment by ricardoV94** [comment]
At least we are narrowing down on what the blockers are!

**Comment by maresb** [comment]
I triggered a bot-rerun of the mkl-service feedstock, so hopefully in some hours there will be some movement on that front. The other possibility would be user-managed BLAS, but that feels like a pretty drastic change.

**Comment by maresb** [comment]
I fixed mkl-service upstream. Let's see if we get further now.

**Comment by maresb** [comment]
Now we are hitting https://github.com/numba/numba/issues/9197. The Numba release canddate is under review, so hopefully it will get released soon! (We may want to test.)

**Comment by ricardoV94** [comment]
Numba is not a hard constraint, we can just not test it in the 3.12 run, like we did for a long time with 3.11

Numba releases and "soon" don't usually go together

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Just need to remove the old Check code style (3.10) | (3.11), which we can do after merging the PR. Otherwise ready to merge?

**Comment by maresb** [comment]
Woah, crazy! Looks like I got it!

> Just need to remove the old Check code style (3.10) | (3.11) from the Github branch protection settings

I don't understand :point_up: but probably I don't need to.

All good from my side! :grin:

**Comment by ricardoV94** [comment]
I was referring to the "required" CI tests that no longer exist: 
image

**Comment by ricardoV94** [comment]
@maresb do you want me to squash and merge or rebase and merge?

**Comment by ricardoV94** [comment]
> I feel like my commits are reasonably structured, so assuming you don't have any criticisms I'd prefer a rebase + merge.

Done, I was just double checking!

**Pull request merged**